### ISSUE TITLE
Update export_kernel_fpu_functions.patch for 5.2?

### ISSUE THREAD

**Issue by fryfrog** [NixOS/nixpkgs#64613]
Hi, I <3 your `export_kernel_fpu_functions.patch` to bring SIMD/FPU back to recent kernels and use it for my Arch Linux kernel. It looks like they've changed enough that still applies, but with some fuzz/offset.

```
patching file arch/x86/include/asm/fpu/api.h
Hunk #1 succeeded at 19 (offset 1 line).
patching file arch/x86/kernel/fpu/core.c
Hunk #1 succeeded at 94 (offset 1 line).
Hunk #2 succeeded at 114 with fuzz 2 (offset 3 lines).
Hunk #3 succeeded at 120 with fuzz 1 (offset -2 lines).
```

Any plans to update it for 5.2 kernels? :)

https://raw.githubusercontent.com/NixOS/nixpkgs/master/pkgs/os-specific/linux/kernel/export_kernel_fpu_functions.patch

https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/kernel/export_kernel_fpu_functions.patch

```
diff --git a/arch/x86/include/asm/fpu/api.h b/arch/x86/include/asm/fpu/api.h
index b774c52e5411..de0d4b67fde7 100644
--- a/arch/x86/include/asm/fpu/api.h
+++ b/arch/x86/include/asm/fpu/api.h
@@ -19,6 +19,8 @@
  * If you intend to use the FPU in softirq you need to check first with
  * irq_fpu_usable() if it is possible.
  */
+extern void __kernel_fpu_begin(void);
+extern void __kernel_fpu_end(void);
 extern void kernel_fpu_begin(void);
 extern void kernel_fpu_end(void);
 extern bool irq_fpu_usable(void);
diff --git a/arch/x86/kernel/fpu/core.c b/arch/x86/kernel/fpu/core.c
index 649fbc3fcf9f..76dfed2bf696 100644
--- a/arch/x86/kernel/fpu/core.c
+++ b/arch/x86/kernel/fpu/core.c
@@ -94,7 +94,7 @@ bool irq_fpu_usable(void)
 }
 EXPORT_SYMBOL(irq_fpu_usable);

-static void __kernel_fpu_begin(void)
+void __kernel_fpu_begin(void)
 {
        struct fpu *fpu = &current->thread.fpu;

@@ -114,11 +114,13 @@ static void __kernel_fpu_begin(void)
        }
        __cpu_invalidate_fpregs_state();
 }
+EXPORT_SYMBOL(__kernel_fpu_begin);

-static void __kernel_fpu_end(void)
+void __kernel_fpu_end(void)
 {
        kernel_fpu_enable();
 }
+EXPORT_SYMBOL(__kernel_fpu_end);

 void kernel_fpu_begin(void)
 {
```

^ That is my `git diff` after applying the patch.

**Comment by fryfrog** [comment]
It fails to apply to 5.3 kernel :(

```
patching file arch/x86/include/asm/fpu/api.h
Hunk #1 succeeded at 19 (offset 1 line).
patching file arch/x86/kernel/fpu/core.c
Hunk #1 FAILED at 93.
Hunk #2 FAILED at 111.
Hunk #3 succeeded at 134 with fuzz 2 (offset 13 lines).
2 out of 3 hunks FAILED -- saving rejects to file arch/x86/kernel/fpu/core.c.rej
```

**Comment by fryfrog** [comment]
But it looks like 0.8.2 is going to include the ZFS work arounds for SIMD so not really needed.

**Comment by fryfrog** [comment]
Nooooo, 0.8.2 pulled the work-a-round because it caused problems in older kernels. See https://github.com/zfsonlinux/zfs/issues/9346 for details.

**Comment by Mic92** [comment]
Yes. We should update the patch for newer kernels.

**Comment by Mic92** [comment]
We have updated the patch. It is even included in 5.6
